File: kirun-js/src/engine/dsl/parser/ast/SchemaLiteralNode.ts
import { SourceLocation } from '../../lexer/DSLToken';
import { ASTNode } from './ASTNode';
import { ExpressionNode } from './ExpressionNode';
import { SchemaNode } from './SchemaNode';

/**
 * Schema literal node - represents inline schema definitions with default values
 * Example: (ARRAY OF INTEGER) WITH DEFAULT VALUE []
 */
export class SchemaLiteralNode extends ASTNode {
    constructor(
        public schema: SchemaNode,
        public defaultValue: ExpressionNode | undefined,
        location: SourceLocation,
    ) {
        super('SchemaLiteral', location);
    }

    public toJSON(): any {
        return {
            type: this.type,
            schema: this.schema.toJSON(),
            defaultValue: this.defaultValue?.toJSON(),
        };
    }
}

File: kirun-py/src/kirun_py/dsl/parser/ast/schema_literal_node.py
from __future__ import annotations

from typing import Any, Optional

from kirun_py.dsl.lexer.dsl_token import SourceLocation
from kirun_py.dsl.parser.ast.ast_node import ASTNode
from kirun_py.dsl.parser.ast.expression_node import ExpressionNode
from kirun_py.dsl.parser.ast.schema_node import SchemaNode


class SchemaLiteralNode(ASTNode):
    """
    Schema literal node - represents inline schema definitions with default values.
    Example: (ARRAY OF INTEGER) WITH DEFAULT VALUE []
    """

    def __init__(
        self,
        schema: SchemaNode,
        default_value: Optional[ExpressionNode],
        location: SourceLocation,
    ) -> None:
        super().__init__('SchemaLiteral', location)
        self.schema = schema
        self.default_value = default_value

    def to_json(self) -> Any:
        result: dict = {
            'type': self.type,
            'schema': self.schema.to_json(),
        }
        if self.default_value is not None:
            result['defaultValue'] = self.default_value.to_json()
        else:
            result['defaultValue'] = None
        return result
